pkg/apis/ceph.rook.io/v1/network_test.go

```diff
@@ -72,10 +72,54 @@ func TestValidateNetworkSpec(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+// test the NetworkSpec.IsHost method with different network providers
+func TestNetworkCephIsHost(t *testing.T) {
+	net := NetworkSpec{HostNetwork: false}
+
+	net.Provider = NetworkProviderHost
+	assert.True(t, net.IsHost())
+
+	net.Provider = NetworkProviderMultus
+	assert.False(t, net.IsHost())
+
+	// test with  nonempty but invalid provider
+	net.Provider = "foo"
+	assert.False(t, net.IsHost())
+
+	// test with  nonempty but invalid provider
+	net.Provider = "foo"
+	assert.False(t, net.IsHost())
+}
+
+// test the NetworkSpec.IsHost method with network providers in combination with the legacy
+// "HostNetwork  setting.
 func TestNetworkCephIsHostLegacy(t *testing.T) {
 	net := NetworkSpec{HostNetwork: true}
 
 	assert.True(t, net.IsHost())
+
+	net.Provider = NetworkProviderDefault
+	assert.True(t, net.IsHost())
+
+	net.Provider = NetworkProviderMultus
+	assert.False(t, net.IsHost())
+
+	net.HostNetwork = false
+	net.Provider = NetworkProviderDefault
+	assert.False(t, net.IsHost())
+
+	net.HostNetwork = true
+	net.Provider = NetworkProviderHost
+	assert.True(t, net.IsHost())
+	net.HostNetwork = true
+	// test with  nonempty but invalid provider
+	net.Provider = "foo"
+	assert.False(t, net.IsHost())
+
+	net.HostNetwork = true
+	// test with  nonempty but invalid provider
+	net.Provider = "foo"
+	assert.False(t, net.IsHost())
```

Comment:
These 2 test cases are the same I think?

-----

Also, bigger picture here: the thing I am most concerned about is that this is really an error condition that Rook shouldn't accept as an input. Setting `hostNetwork: true` with any network provider other than `provider: host` is ambiguous and is therefore an error condition.

That said, I think this behavior is reasonable. It's okay for this function to not have an `err` return so that we don't have to change all instances of it. And it is good to test all expected conditions in CI.

However, we also want to make sure that the operator won't accept this during runtime. Please document this test condition stating that we are intentionally not returning an error because we rely on the operator to catch the error condition before this function is used.

How do you feel about adding a CephCluster controller test that ensures this code is called? (And adding comments to it to mention that the test is critical to ensure `IsHost()` is valid.) https://github.com/rook/rook/blob/4ed35d6bd671883172f3f4bfd3ec5721ff6f3f3d/pkg/operator/ceph/cluster/cluster.go#L299-L301

I think that test would happen in this file: https://github.com/rook/rook/blob/f007f2aca1f8d7bfbd59968e14b64738897cd67d/pkg/operator/ceph/cluster/cluster_test.go

Reply:
@BlaineEXE  wrote:
> These 2 test cases are the same I think?

right. I had already removed the duplicate a little earlier today.


 
> Also, bigger picture here: the thing I am most concerned about is that this is really an error condition that Rook shouldn't accept as an input. Setting `hostNetwork: true` with any network provider other than `provider: host` is ambiguous and is therefore an error condition.
> 
> That said, I think this behavior is reasonable. It's okay for this function to not have an `err` return so that we don't have to change all instances of it. And it is good to test all expected conditions in CI.
> 
> However, if we do this, we also want to make sure that the operator won't accept this during runtime. Please document this test condition stating that we are intentionally not returning an error because we rely on the operator to catch the error condition before this function is used.
> 
> How do you feel about adding a CephCluster controller test that ensures this code is called? (And adding comments to it to mention that the test is critical to ensure `IsHost()` is valid.)
> 
> https://github.com/rook/rook/blob/4ed35d6bd671883172f3f4bfd3ec5721ff6f3f3d/pkg/operator/ceph/cluster/cluster.go#L299-L301
> 
> I think that test would happen in this file: https://github.com/rook/rook/blob/f007f2aca1f8d7bfbd59968e14b64738897cd67d/pkg/operator/ceph/cluster/cluster_test.go
> 
> **Update:** The alternative is to make `IsHost()` return an error, which is the bigger-picture easier thing to do IMO.

> What do you think?

I agree. and I tend to make `IsHost` validate the networkSpec and possibly return error.

I had actually considered doing this originally, but now I am not so sure anymore, why I ended up not doing it.
Anyway, I will change the patch accordingly ;)